pkg/proxy/iptables/proxier.go

```diff
@@ -673,44 +686,49 @@ func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
 func (proxier *Proxier) OnNodeSynced() {
 }
 
+//FIXME:
 // portProtoHash takes the ServicePortName and protocol for a service
 // returns the associated 16 character hash. This is computed by hashing (sha256)
 // then encoding to base32 and truncating to 16 chars. We do this because IPTables
 // Chain Names must be <= 28 chars long, and the longer they are the harder they are to read.
-func portProtoHash(servicePortName string, protocol string) string {
-	hash := sha256.Sum256([]byte(servicePortName + protocol))
+type iptablesHasher struct{}
+
+func (iptablesHasher) Hash(input ...string) string {
+	buf := bytes.Buffer{}
+	for _, in := range input {
+		buf.WriteString(in)
+	}
+	hash := sha256.Sum256(buf.Bytes())
 	encoded := base32.StdEncoding.EncodeToString(hash[:])
 	return encoded[:16]
 }
 
 // servicePortChainName takes the ServicePortName for a service and
 // returns the associated iptables chain.  This is computed by hashing (sha256)
 // then encoding to base32 and truncating with the prefix "KUBE-SVC-".
-func servicePortChainName(servicePortName string, protocol string) utiliptables.Chain {
-	return utiliptables.Chain("KUBE-SVC-" + portProtoHash(servicePortName, protocol))
+func servicePortChainName(hasher hasher, servicePortName string, protocol string) utiliptables.Chain {
+	return utiliptables.Chain("KUBE-SVC-" + hasher.Hash(servicePortName, protocol))
```

Comment:
due to the aforementioned caching, each of these functions is only called from a single place... maybe get rid of the extra functions, and have:

```
type iptablesChainFunc func(prefix string, ids ...string) utiliptables.Chain

func defaultIPTablesChainMaker(prefix string, ids ...string) utiliptables.Chain {
	buf := bytes.Buffer{}
	for _, id := range ids {
		buf.WriteString(id)
	}
	hash := sha256.Sum256(buf.Bytes())
	encoded := base32.StdEncoding.EncodeToString(hash[:])
	return utiliptables.Chain(prefix+encoded[:16])
}

func newServiceInfo(makeIPTablesChain iptablesChainFunc) func(port *v1.ServicePort, service *v1.Service, baseInfo *proxy.BaseServiceInfo) proxy.ServicePort {
		...
		info.servicePortChainName = makeIPTablesChain("KUBE-SVC-", info.serviceNameString, protocol)
```